src/groupeddataframe/splitapplycombine.jl

```diff
@@ -962,14 +1056,51 @@ function _combine(f::AbstractVector{<:Pair},
                 throw(ArgumentError("a single value or vector result is required when passing " *
                                     "multiple functions (got $(typeof(res)))"))
             end
+            if !(firstres isa AbstractVector) && isnothing(idx_agg)
+                idx_agg = Vector{Int}(undef, length(gd))
+                fillfirst!(nothing, idx_agg, 1:length(gd.groups), gd)
+            end
             idx, outcols, _ = _combine_with_first(wrap(firstres), fun, gd, incols,
-                                                  Val(firstmulticol))
+                                                  Val(firstmulticol),
+                                                  firstres isa AbstractVector ? nothing : idx_agg)
+            @assert length(outcols) == 1
             res[i] = idx, outcols[1]
         end
     end
-    # TODO: avoid recomputing idx for each pair when !all(isagg, f)
-    idx = res[1][1]
+    idx_loc = findfirst(x -> x[1] !== idx_agg, res)
+    if isnothing(idx_loc)
+        @assert !isnothing(idx_agg)
+        idx = idx_agg
+    else
+        idx = res[idx_loc][1]
+        agg2idx_map = nothing
+        for i in 1:length(res)
+            if res[i][1] != idx
+                if res[i][1] === idx_agg
```

Comment:
Why not do this out of the `for` loop instead of doing `agg2idx_map = nothing`? AFAICT you're certain that `res[i][1] === idx_agg` will be true for at least one `i`, right?

Reply:
No, `idx_agg` will be `nothing` if no columns were of one-row type (i.e. all transformation functions returned vectors). I use `nothing` for `agg2idx_map` as a sentinel that no column of this type was seen yet. If it is encountered for the first time I populate it.

I will leave this open as this is a second crucial part of the PR (i.e. that I recompute `idx_agg` and `agg2idx_map` at most once and only if at least one function producing single rows were passed by the user).